developer-docs-site/docs/tutorials/first-dapp.md

```diff
@@ -368,6 +368,217 @@ You can also verify that the module was published by going to the [Aptos Explore
+cat package-metadata.bcs | od -v -t x1 -A n | tr -d ' \n'
+```
+That will output a hex string we can later use.
+
+Convert `message.mv` to a hex string. On Mac and Linux we can use the command
+```bash
+cat bytecode_modules/message.mv | od -v -t x1 -A n | tr -d ' \n'
+```
+That will output a hex string we can later use.
+
+Keep both of the hex strings so later we can use them!
+
+Back to our react app, let's add a button to click on to publish the module, use the `publishPackage` function TS SDK provides us and display a link to get the account's resources where we can see the published module.
+
+We would need our account's private key (you can get the private key from the Petra Wallet by going to settings, manage account, show the private key, and copy that field) to initialize an `AptosAccount` to publish the module with. Since a private key is a VERY sensitive data, we dont want to expose it in the code, but to hold it in an `.env` file and use it from there.
+
+1. Create a new `.env` file on the `root` of the project and add to the file
+```bash
+REACT_APP_ACCOUNT_PK=<account-private-key>
+```
+Make sure to restart the local server so the app will load the new `.env` file.
+
+2. Add the following to `src/App.tsx`, where 
+- `process.env.REACT_APP_ACCOUNT_PK` holds the account private key. 
+- `<package-metadata.bcs hex string>` is the `package-metadata.bcs` hex string output we get from the previous step
+- `<message.mv hex string>` is the `message.mv` hex string output we get from the previous step
+
+```typescript
+import { Types, AptosClient, AptosAccount, HexString, TxnBuilderTypes} from "aptos";
+  // ...
+
+function App() {
+  // ...
+
+  // Publish the module using the TS SDK
+  const [publishPackageTxnHash, setPublishPackageTxnHash] = useState<string | null>(null);
+  const [isPublishing, setIsPublishing] = useState<boolean>(false);
+  const onPublishModule = async () => {
+    if (!process.env.REACT_APP_ACCOUNT_PK) return;
+    setIsPublishing(true);
+    const aptosAccount = new AptosAccount(
+      new HexString(process.env.REACT_APP_ACCOUNT_PK).toUint8Array()
+    );
+    const txnHash = await client.publishPackage(
+      aptosAccount,
+      new HexString(
+        // package-metadata
+        "<package-metadata.bcs hex string>"
+      ).toUint8Array(),
+      [
+        new TxnBuilderTypes.Module(
+          new HexString(
+            // modules
+            "<message.mv hex string>"
+          ).toUint8Array()
+        ),
+      ]
+    );
+
+    await client.waitForTransaction(txnHash);
+    setIsPublishing(false);
+    setPublishPackageTxnHash(txnHash);
+  };
+
+  return (
+    <div className="App">
+      // ...
+      <div>
+        <button onClick={onPublishModule} disabled={isPublishing}>
+          Publish Package
+        </button>
+        {publishPackageTxnHash && (
+          <div>
+            <p>
+              <a
+                href={`https://fullnode.devnet.aptoslabs.com/v1/accounts/${address}/modules`}
```

Comment:
Other thing is we probably need to check if this is all done on devnet / ensure it's consistent.  Last time I looked I think it was devnet, but just to double check

Reply:
Agreed, i think it does use devnet. maybe we should decide to use devnet only in this tutorial and point links/explorer to devnet